[diff_drive_controller] Add wheel_separation and wheel_radius parameters

bmagyar opened the pull request:
@T045T, @efernandez

bmagyar committed:
If wheel separation and wheel diameter is specified, don't look them up from urdf

bmagyar commented:
@T045T, you have push access to the branch of this PR in my fork. :+1:

T045T committed:
allow the wheel separation and radius to be set from different sources

i.e. one can be set from the URDF, the other from a parameter

T045T committed:
add a sphere-wheeled version of diffbot to test the new wheel_radius parameter

T045T committed:
add test that checks that controller initialization fails when wheels are not cylinders and no wheel_radius parameter is given

T045T committed:
add test that checks wheel_radius parameter functionality

T045T committed:
add test for wheel_separation parameter

this doesn't technically test whether the separation is set correctly, but observing the controller's info output shows that it is

T045T commented:
@bmagyar I've written some tests, and rewritten the code so it's possible to specify just the separation or just the radius as a parameter. Can you review these changes?

bmagyar commented on diff_drive_controller/src/diff_drive_controller.cpp:
```diff
@@ -193,8 +193,31 @@ namespace diff_drive_controller{
+    bool lookup_wheel_radius = !controller_nh.getParam("wheel_radius", wheel_radius_);
+
+    // Short-circuit in case we don't need to look up anything, so we don't have to parse the URDF
+    if(lookup_wheel_separation or lookup_wheel_radius)
```
Excellent idea, but then why not short-circuit in the function? This would move the bloat a bit away from the relatively clean param parsing chunk of code.

T045T replied:
Good point, I'll change that. Also, since `setOdomParamsFromUrdf` doesn't actually set the parameters any more, is it okay to rename it to `lookupOdomParamsInUrdf`?

bmagyar replied:
I would say it does.
https://github.com/bmagyar/ros_controllers/blob/separation_and_diameter_params/diff_drive_controller/src/diff_drive_controller.cpp#L476

T045T replied:
But the actual odometry values are now being set [here](https://github.com/bmagyar/ros_controllers/blob/separation_and_diameter_params/diff_drive_controller/src/diff_drive_controller.cpp#L217). This is obviously a minor point, I just want to make sure the function name doesn't confuse future developers :)

T045T committed:
move short-circuit of URDF lookup into setOdomParamsFromUrdf

bmagyar commented on diff_drive_controller/src/diff_drive_controller.cpp:
```diff
@@ -407,8 +427,16 @@ namespace diff_drive_controller{
+                             bool lookup_wheel_separation,
+                             bool lookup_wheel_radius)
   {
+    if(not(lookup_wheel_separation or lookup_wheel_radius))
```
Even though I think these keywords are awesome, the rest of the project is not using them, so I'd ask you to remove them to keep the consistency of `ros_controllers`

bmagyar commented on diff_drive_controller/src/diff_drive_controller.cpp:
```diff
@@ -193,8 +193,28 @@ namespace diff_drive_controller{
+    bool lookup_wheel_separation = !controller_nh.getParam("wheel_separation", wheel_separation_);
+    bool lookup_wheel_radius = !controller_nh.getParam("wheel_radius", wheel_radius_);
+
+
```
You have 2 empty lines here, one would be enough.

bmagyar commented on diff_drive_controller/src/diff_drive_controller.cpp:
```diff
@@ -421,46 +449,48 @@ namespace diff_drive_controller{
-    }
+      // Get wheel separation
+      if(!left_wheel_joint)
+        {
```
A little bit of misindentation here

T045T committed:
style and whitespace fixes

T045T commented:
Thanks! I've fixed those three :)

bmagyar commented:
Nice work!
I made small pedantic notes and I ran all the tests and they are fine.

But... I have spotted that the urdf you added with spherical wheels is actually almost the same as this one:
https://github.com/bmagyar/ros_controllers/blob/indigo-devel/diff_drive_controller/test/diffbot_bad.xacro
At that moment this was considered a "bad" urdf, but not I believe it could be used without any issues. Actually the one you added has a more proper name. 
There is one test that checks for a non-standard urdf, so your test doesn't need to validate the "no param added and urdf is not the expected structure" case, since there is a test for that already. But! it would be good to make it more clear what bad urdf means in those cases.

T045T commented:
I looked at the bad URDF test as well, and in addition to the spherical wheels, it also has the wrong joint names.
The [controller config](https://github.com/bmagyar/ros_controllers/blob/indigo-devel/diff_drive_controller/test/diffbot_controllers.yaml) uses `wheel_0_joint` and `wheel_1_joint`, while the URDF defines `joint_w1` and `joint_w2`, so it can't be used for a passing test, sadly.

Looks like the wront joint names weren't actually what that test was supposed to check though (`diffbot_drive_wrong.test` does that), so maybe we can just remove the outdated "bad" URDF outright and replace `diffbot_drive_bad_urdf.test` with `diffbot_drive_radius_param_fail.test`?

bmagyar commented:
Good catch!
Give it a shot and if all unit tests are still ok, we should be done.

efernandez commented:
Nice work guys... I don't have the cycles to check this now, but it seems OK at first sight.

T045T committed:
remove diff_drive_bad_urdf.test

The 'bad' URDF is not that bad any more since the wheel radius can be set via parameter for non-cylindrical wheels.
However, it used outdated joint names that would cause controller creation to fail even with the wheel_radius parameter set.
Since diff_drive_wrong.test already checks that case, and diff_drive_radius_param_fail.test explicitly checks for behavior with missing wheel_radius, diff_drive_bad_urdf.test is obsolete

paulbovbel commented:
+1, tested with https://github.com/clearpathrobotics/urdf_tutorial/tree/ros-control-tutorials, works as expected

bmagyar commented:
+1 from me too.

@adolfo-rt ?

adolfo-rt commented:
Please update the package's ROS wiki page, so users can know about the new configuration alternatives.

Thanks for the patch, and for increasing test coverage :)

paulbovbel commented:
Would you please do a release with these changes, so that I can work diff_drive_controller into http://wiki.ros.org/urdf/Tutorials

T045T commented:
I've updated http://wiki.ros.org/diff_drive_controller with the new parameters.

bmagyar commented:
Awesome, thanks Nils!
